Bids utilities

MartinPerez opened the pull request:
This PR introduces the first_level_models_from_bids object that allows to automatically extract relevant files by just providing a bids directory. This goes along what seem to be stable agreements on the bids derivatives document still under discussion. Particular on how to name preprocessed bold files and other simple problems. Moreover a downloader for all openfmri datasets is provided alongside a downloader for a downsampled language localizer dataset in the bids format on top of which an example is built for the documentation. All tests and documentation are complete.

The only thing missing is to host the resampled language localizer dataset in the cea server and to first merge the sphinx PR on which this PR depends. Then it will just be a matter of a quick update for this PR.

MartinPerez commented:
Introduced minimal change on the API, changing subject_id in FirstLevelModel to subject_label to comply correctly with BIDS documentation. And updating everywhere else as necessary.

MartinPerez commented:
There are a couple of nice additions on utilities employed by the rest of the code to simplify searching groups of files in a bids dataset without knowledge of all the underlying conventions, and to parse the file names into a dictionary of field-key pairs.

bthirion commented:
If I understand correctly, this PR includes the sphinx_docs one ( #97  ). I think we should first merge #97 to make the change list less cluttered.

MartinPerez commented:
Yes that is necessary before merging this. Also we need to host the resampled bids dataset.

bthirion commented:
By 'resampled', you mean: put into BIDS layout ?
I don't see that as a blocker though: Nistats   should also work with non-BIDS datasets ?

MartinPerez commented:
By resampled I mean the images were resampled to be light, its a light BIDS dataset with the language localizer experimental paradigm for 10 subjects (around 100MB per subject, we could make it even lighter, less resolution/less subjects). Its necessary for the BIDS example I created, since the others are monster datasets that take hours to download.

bthirion commented:
OK. I think that it should be downloaded to OpenfMRI then.

MartinPerez commented:
If they are willing that would be nice. But its kind of a dummy dataset, I mean it only has what we need in the raw and derivatives folder to run our analysis. The standard to be uploaded to OpenFmri might require a lot of other documentation and clarifications that we dont need for the example alongside the raw bold images that im not including to keep it light, since we work on the preprocessed images only. Which is why I was hoping we could host this simplified version ourselves, like the nipy datasets. Shall I send an email to Chris Gorgolewsky to ask him about it?

MartinPerez commented:
@bthirion I rebased, all tests passed, except for not having hosted the dataset of the bids example.

MartinPerez commented:
All tests passed and the bids dataset is hosted. @bthirion I managed to compress the dataset to 270 MB without losing too much on the resulting contrast images. I await review.

bthirion commented on nistats/utils.py:
```diff
@@ -204,3 +206,159 @@ def pos_recipr(X):
+        In case filters are provided, this option detemines if the filters
+        provided for the search are supposed to be only fields in the file
+        name. If set to False then any file name that contains additional
+        fields will be removed from the results list.
```
I don't find this very clear. Can you explain the use case  / give an example ?

MartinPerez replied:
Its intended for cases in which one would want to confirm the filters considered accounted for all the fields in the file names. Also when uniqueness of files wants to be enforced (since many files like variants and spaces not specified by the filters might be available), maybe I should handle the uniqueness by throwing an error instead? I used this many times as sanity check when exploring the content of a dataset. It seemed useful at some point, but maybe is too ill defined?

MartinPerez replied:
yeah looking closer it looks like feature creeping, I am taking it out.

bthirion commented on nistats/utils.py:
```diff
@@ -204,3 +206,159 @@ def pos_recipr(X):
+
+    Returns
+    -------
+    files: list of str or list of dict
```
I find this a bit annoying. When should this be a list of dict ?

bthirion commented on nistats/utils.py:
```diff
@@ -204,3 +206,159 @@ def pos_recipr(X):
+    reference['file_path'] = img_path
+    reference['file_basename'] = os.path.basename(img_path)
+    parts = reference['file_basename'].split('_')
+    tag, typ = parts[-1].split('.', 1)
```
type

bthirion commented on nistats/first_level_model.py:
```diff
@@ -557,3 +568,214 @@ def compute_contrast(self, contrast_def, stat_type=None,
+        different sessions are considered together for the same subject to run
+        a fixed effects analysis on them.
+
+    models_fit_kwargs: list of dict
```
I thing that this overly generic argument is bad design. Imho we need to enforce much more structure in it.

MartinPerez replied:
Ok I will change it

bthirion commented on nistats/first_level_model.py:
```diff
@@ -557,3 +568,214 @@ def compute_contrast(self, contrast_def, stat_type=None,
+        and a confounds file, so all three lists must have the same size.
+        fit_kwargs are provided as a dictionary to be easily passed to the
+        fit function of the models as model.fit(**kwargs), particularly when
+        no additional modifications of events or confounds is necessary.
```
But then  it can go wrong arbitrary poorly I'm afraid.Why not simply returning run_imgs, events and optionally confounds ?

MartinPerez replied:
I think returning optionally arguments like confounds is confusing for users, I would return a None value for it instead which is coherent with the FirstLevelModel fit API.

bthirion commented:
This PR is really great. My main suggestion would be to split it to make tracking a bit faster (e.g. by taking care of the dowloader in another PR). Since you did that way, let's keep it as is.
Otherwise, my main comment is to avoid dictionaries as arguments, which makes  the API esoteric.

MartinPerez commented on examples/plot_bids_analysis.py:
```diff
@@ -0,0 +1,114 @@
+# since in this case a confounds.tsv file is available in the BIDS dataset.
+# To get the first level models we only have to specify the dataset directory
+# and the task_id as specified in the file names.
+task_id = 'languagelocalizer'
```
not if we follow BIDS specification. The specification says: "Each task has a unique label consisting of letters and/or numbers (other characters including spaces and underscores are not allowed)."

bthirion replied:
Then it should be LanguageLocalizer.

MartinPerez commented on examples/plot_second_level_design_matrix.py:
```diff
@@ -21,18 +21,18 @@
 # --------------------------------------
 # Experimental paradigm has two conditions and 20 subjects
-column_names = ['map_name', 'subject_id', 'map_path']
+column_names = ['map_name', 'subject_label', 'map_path']
```
Yes, to follow more closely the BIDS specification. ids refer to the whole string in the filenames like "sub-01" while label would refer only to "01". So here we are suppose to pass the subject_label and not the subject_id.

MartinPerez commented on nistats/first_level_model.py:
```diff
@@ -444,8 +451,12 @@ def fit(self, run_imgs, events=None, confounds=None,
-            Y = self.masker_.transform(run_img)
+                sys.stderr.write('Starting masker computation \r')
+
+            # ATTENTION THIS IS NOT SUPPOSED TO BE HERE
```
Oh definitely im sorry, I forgot to swap back this line of code. It was a simple sanity check on the masker behavior but I forgot to revert it, good thing I commented on that. Also the import showed the same comment.

MartinPerez commented on examples/plot_bids_analysis.py:
```diff
@@ -0,0 +1,114 @@
+                              plot_abs=False, display_mode='x')
+fig.suptitle('subjects z_map language netowrk (language - string)')
+plt.show()
+
```
Definitely for the group results map, but adding the colorbar for each small brain looked visually unappealing when I tried. Anyway I show all the brains just for illustrative purposes to compare visually the patterns with the group result. An option would be to make a common colorbar with the max stat of all images, but seems to me to complicate the code in the example without adding much value to it.

MartinPerez commented on nistats/first_level_model.py:
```diff
@@ -557,3 +568,214 @@ def compute_contrast(self, contrast_def, stat_type=None,
+    task_id: str
+        Task_id as specified in the file names. Of the form _task-*_.
+
+    model_init: dict, optional
```
The dictionary is not part of BIDS. I would say this is more an utility to avoid modifying one by one the returned models with different parameters from the default like I was doing initially. I chose a dictionary for two reasons. 1) The parameters are very numerous and irrelevant to the notion of getting the models from BIDS. 2) To avoid the problem of maintaining this function arguments if the FirstLevelModel arguments change in any way.

MartinPerez replied:
It could also be useful just to load a model set of parameters saved as a dictionary in a json file for a Nistats model template.

bthirion replied:
As long as the BIDS norm does not say anything, I'd be in favor of explicit arguments. It is very likely that BIDS will come with a specific for these arguments, and then we'll use the API they suggest (e.g. a json file).
But I really want to avoid unconstrained dictionaries, because it is impossible to assert whether the argument is acceptable or not.

MartinPerez commented on nistats/first_level_model.py:
```diff
@@ -557,3 +568,214 @@ def compute_contrast(self, contrast_def, stat_type=None,
+    # Only if not given in model_init dictionary
+    # Throw warning if no bold.json is found
+    TR = None
+    SliceTimingRef = 0.0
```
I agree, this depends on accepting or not to have model_init as the function argument, if instead we specify all firstlevelmodel parameters this would be there.

MartinPerez commented on nistats/first_level_model.py:
```diff
@@ -557,3 +568,214 @@ def compute_contrast(self, contrast_def, stat_type=None,
+                              file_tag='preproc', file_type='nii*',
+                              sub_label=sub_label, filters=filters)
+        # If there is more than one file for the same (ses, run), likely we
+        # have an issue of underspecification of filters.
```
What do you mean by the _RL and _LR data? this would capture *preproc.nii files. Then, the idea I had here is that if there are many variants and spaces, since we can not really infer a default, I throw an error.

MartinPerez commented on nistats/utils.py:
```diff
@@ -204,3 +206,159 @@ def pos_recipr(X):
+        at the level of directories. the label is what follows the 'sub' field
+        in the BIDS convention as 'sub-label'.
+
+    file_folder: str accepted by glob, optional (default: '*')
```
it could also be a user provided folder with arbitrary meaning, it was difficult to name so I settled for a general name (name of the folder that contains the file), but I am also not completely convinced by it.

bthirion replied:
If you don't have any more precise use case in mind, I would change it to 'modality', to keep these fields explicit. All these functions abstract information, and I would like parameters to be as concrete as possible.
But if you have another possibility in mind, state it.
In general, I think that docstrings should not be too asbtract, but provide examples whenever possible.

MartinPerez commented:
I changed the bids api to return properly lists of relevant arguments instead of a dictionary.

I also simplified get_bids_files to avoid returning dictionaries, it was a bad decision carried forward after refactoring parse_bids_files as an independent function, I see it can simply be called afterwards if desired. I also got rid of allow_other_fields, it seemed like feature creeping at second inspection.

I addressed or responded to all other comments. I only have doubts about changing the model_init argument of the first_level_models_from_bids function

bthirion commented:
On 05/02/2017 01:57, Martin Perez-Guevara wrote:
>
> *@MartinPerez* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In examples/plot_bids_analysis.py 
> <https://github.com/nistats/nistats/pull/104>:
>
> > +# information to run an statistical analysis using Nistats. The raw data
> +# subject folders only contain bold.json and events.tsv files, while the
> +# derivatives folder with preprocessed files contain preproc.nii and
> +# confounds.tsv files.
> +data_dir, _ = fetch_bids_langloc_dataset()
> +
> +#############################################################################
> +# Obtain automatically FirstLevelModel objects and fit arguments
> +# --------------------------------------------------------------
> +# From the dataset directory we obtain automatically FirstLevelModel objects
> +# with their subject_id filled from the BIDS dataset. Moreover we obtain
> +# for each model a dictionary with run_imgs, events and confounder regressors
> +# since in this case a confounds.tsv file is available in the BIDS dataset.
> +# To get the first level models we only have to specify the dataset directory
> +# and the task_id as specified in the file names.
> +task_id = 'languagelocalizer'
>
> not if we follow BIDS specification. The specification says: "Each 
> task has a unique label consisting of letters and/or numbers (other 
> characters including spaces and underscores are not allowed)."
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/nistats/nistats/pull/104>, or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AAOT1hAuhHUN97u44Rfayo51q54bAfsRks5rZR51gaJpZM4LWiXM>.
>
Yes, I told them it was bad design, but too late.

Then it should be at least LanguageLocalizer

MartinPerez commented:
Made all arguments explicit in FLM from BIDS. Left 'languagelocalizer' as it is until I can swap the hosted dataset, since it is like that in all the file names, I would vote to take care of this detail later. All issues are addressed. Ready to merge once tests pass.

MartinPerez commented:
@lesteve could you please take a quick look and tell me if you have encountered before a similar failure of sphinx-gallery in circle ci? locally I run the examples without problems and the error message is quite vague :S. Also it was running fine one commit before without touching the examples :S.

lesteve commented:
Looks like a genuine error to me:

```
Traceback (most recent call last):
  File "/home/ubuntu/nistats/doc/sphinxext/sphinx_gallery/gen_rst.py", line 518, in execute_code_block
    exec(code_block, example_globals)
  File "<string>", line 6, in <module>
IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices
```

Try using the same numpy version as the one on CircleCI (I am guessing this is numpy 1.12 but you can double-check from the log).

Using a more recent of sphinx-gallery you probably would have got a better error message, just saying ;-).

MartinPerez commented:
Thanks @lesteve, this thursday I will PR the new sphinx gallery, it was indeed a problem with the last numpy. @bthirion all tests passed, ready for final review/ merge.

bthirion commented:
OK, I'll take a look asap --which is unfortunately not immediately...

bthirion commented on nistats/tests/test_utils.py:
```diff
@@ -100,3 +105,138 @@ def test_img_table_checks():
+
+
+def basic_confounds(length):
+    columns = ['RotX', 'RotY', 'RotZ', 'X', 'Y', 'Z']
```
['RotX', 'RotY', 'RotZ', 'TrX', 'TrY', 'TrZ'] would be more consistent.

MartinPerez replied:
I took the variable names from the BIDS documentation

bthirion commented on nistats/tests/test_utils.py:
```diff
@@ -100,3 +105,138 @@ def test_img_table_checks():
+    return file_path
+
+
+def basic_paradigm():
```
Ideally, should be reused in test_dmtx

MartinPerez commented on examples/plot_bids_analysis.py:
```diff
@@ -0,0 +1,114 @@
+# We just have to provide the list of fitted FirstLevelModel objects
+# to the SecondLevelModel object for estimation. We can do this since
+# all subjects share the same design matrix.
+first_level_conditions = [['language', 'language'], ['string', 'string']]
```
handled in following PR

MartinPerez commented on examples/plot_bids_analysis.py:
```diff
@@ -0,0 +1,114 @@
+# to the SecondLevelModel object for estimation. We can do this since
+# all subjects share the same design matrix.
+first_level_conditions = [['language', 'language'], ['string', 'string']]
+second_level_input = models
```
handled in following PR
